plugins/processors/topk/topk.go

```diff
@@ -0,0 +1,402 @@
+package topk
+
+import (
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/processors"
+)
+
+type TopK struct {
+	Period            int
+	K                 int
+	GroupBy           []string `toml:"group_by"`
+	GroupByMetricName bool     `toml:"group_by_metric_name"`
+	Fields            []string
+	Aggregation       string
+	Bottomk           bool
+	SimpleTopk        bool   `toml:"simple_topk"`
+	DropNoGroup       bool   `toml:"drop_no_group"`
+	DropNonTop        bool   `toml:"drop_non_top"`
+	GroupByTag        string `toml:"group_by_tag"`
+	PositionField     string `toml:"position_field"`
+	AggregationField  string `toml:"aggregation_field"`
+
+	cache           map[string][]telegraf.Metric
+	metricRegex     *regexp.Regexp
+	tagsRegexes     map[string]*regexp.Regexp
+	lastAggregation time.Time
+}
+
+func NewTopK() TopK {
+	// Create object
+	topk := TopK{}
+
+	// Setup defaults
+	topk.Period = 10
+	topk.K = 10
+	topk.Fields = []string{"value"}
+	topk.Aggregation = "avg"
+	topk.GroupBy = []string{}
+	topk.GroupByMetricName = false
+	topk.GroupByTag = ""
+	topk.SimpleTopk = false
+	topk.DropNoGroup = true
+	topk.DropNonTop = true
+	topk.PositionField = ""
+	topk.AggregationField = ""
+
+	// Initialize cache
+	topk.Reset()
+
+	return topk
+}
+
+func NewTopKProcessor() telegraf.Processor {
+	topk := NewTopK()
+	return &topk
+}
+
+var sampleConfig = `
+[[processors.topk]]
+  period = 10                  # How many seconds between aggregations. Default: 10
+  k = 10                       # How many top metrics to return. Default: 10
+
+  # Metrics are grouped based on their tags and name. The plugin aggregates the selected fields of
+  # these groups of metrics and sorts the groups based these aggregations
+  group_by = ["process_name"]  # Over which tags should the aggregation be done. Default: []
+  group_by_metric_name = false # Wheter or not to also group by metric name. Default: false
+
+  # The plugin can aggregate over several fields. If more than one field is specified, an aggregation is calculated per group per field
+  # The plugin returns a metric if it is in a group in the top k groups ordered by any of the aggregations of the selected fields
+  # This effectively means that more than K metrics may be returned. If you need to return only the top k metrics regardless of grouping, use the simple_topk setting
+  fields = ["memory_rss"]      # Over which fields are the top k are calculated. Default: ["value"]
+  aggregation = "avg"          # What aggregation to use. Default: "avg". Options: sum, avg, min, max
```

Comment:
I wonder if we need to have support for different aggregations per field.

Reply:
That is a very good question. I thought about it and decided that someone down the line could contribute such a feature. Do you think is a valuable enough capability to delay the release and work on it before merging?

Reply:
I think we can add this in later.